File: minigit-4/py/commands/tag.py
"""tag command - Create, list, or delete tags."""

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent))

from utils import find_repo_root, get_minigit_dir
from objects import Tag, write_object, resolve_object
from refs import list_tags, create_tag, delete_tag, resolve_head
from commands.commit import get_author_info


def run(args):
    """Manage tags."""
    repo_root = find_repo_root()
    if repo_root is None:
        print("fatal: not a minigit repository", file=sys.stderr)
        return 1

    # Parse arguments
    annotated = False
    delete = False
    list_mode = False
    message = None
    tag_name = None
    commit_ref = None

    i = 0
    while i < len(args):
        arg = args[i]
        if arg == '-a':
            annotated = True
        elif arg == '-d':
            delete = True
        elif arg == '-l':
            list_mode = True
        elif arg == '-m' and i + 1 < len(args):
            message = args[i + 1]
            annotated = True  # -m implies -a
            i += 1
        elif not arg.startswith('-'):
            if tag_name is None:
                tag_name = arg
            else:
                commit_ref = arg
        i += 1

    # Handle delete
    if delete:
        if not tag_name:
            print("error: tag name required", file=sys.stderr)
            return 1

        if delete_tag(tag_name, repo_root):
            print(f"Deleted tag '{tag_name}'")
            return 0
        else:
            print(f"error: tag '{tag_name}' not found", file=sys.stderr)
            return 1

    # Handle list
    if list_mode or tag_name is None:
        tags = list_tags(repo_root)
        for tag in tags:
            print(tag)
        return 0

    # Create tag
    # Resolve target commit
    if commit_ref:
        try:
            target_sha = resolve_object(commit_ref, repo_root)
        except ValueError as e:
            print(f"fatal: {e}", file=sys.stderr)
            return 1
    else:
        target_sha = resolve_head(repo_root)
        if not target_sha:
            print("fatal: not a valid object name: 'HEAD'", file=sys.stderr)
            return 1

    if annotated:
        if not message:
            print("error: annotated tag requires a message (-m)", file=sys.stderr)
            return 1

        # Create annotated tag object
        tag = Tag()
        tag.object_sha = target_sha
        tag.object_type = 'commit'
        tag.tag_name = tag_name
        tag.tagger = get_author_info()
        tag.message = message

        tag_data = tag.serialize()
        tag_sha = write_object('tag', tag_data, repo_root)

        # Create tag ref pointing to tag object
        if not create_tag(tag_name, tag_sha, repo_root):
            print(f"fatal: tag '{tag_name}' already exists", file=sys.stderr)
            return 1
    else:
        # Create lightweight tag (just a ref)
        if not create_tag(tag_name, target_sha, repo_root):
            print(f"fatal: tag '{tag_name}' already exists", file=sys.stderr)
            return 1

    return 0

File: minigit-4/ts/src/commands/tag.ts
// tag command - Create, list, or delete tags

import { findRepoRoot, getAuthorInfo } from '../utils';
import { getTags, createTag, deleteTag, tagExists, getHeadCommit, resolveRevision } from '../refs';
import { createTagContent, writeObject, readObject, parseTagContent } from '../objects';
import { TagInfo } from '../types';

export function tag(args: string[]): number {
  const repoRoot = findRepoRoot();
  if (!repoRoot) {
    console.error('fatal: not a minigit repository');
    return 1;
  }

  let annotated = false;
  let deleteFlag = false;
  let listFlag = false;
  let message: string | null = null;
  const positional: string[] = [];

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '-a') {
      annotated = true;
    } else if (args[i] === '-d') {
      deleteFlag = true;
    } else if (args[i] === '-l') {
      listFlag = true;
    } else if (args[i] === '-m' && i + 1 < args.length) {
      message = args[++i];
      annotated = true; // -m implies -a
    } else if (!args[i].startsWith('-')) {
      positional.push(args[i]);
    }
  }

  if (deleteFlag) {
    if (positional.length === 0) {
      console.error('fatal: tag name required');
      return 1;
    }
    return deleteTagCmd(repoRoot, positional[0]);
  }

  if (listFlag || positional.length === 0) {
    return listTags(repoRoot);
  }

  // Create tag
  const tagName = positional[0];
  const commitRef = positional[1];

  if (annotated) {
    if (!message) {
      console.error('fatal: -a requires -m <message>');
      return 1;
    }
    return createAnnotatedTag(repoRoot, tagName, commitRef, message);
  }

  return createLightweightTag(repoRoot, tagName, commitRef);
}

function listTags(repoRoot: string): number {
  const tags = getTags(repoRoot);

  for (const t of tags) {
    console.log(t);
  }

  return 0;
}

function createLightweightTag(repoRoot: string, tagName: string, commitRef?: string): number {
  if (tagExists(repoRoot, tagName)) {
    console.error(`fatal: tag '${tagName}' already exists`);
    return 1;
  }

  let sha: string | null;
  if (commitRef) {
    sha = resolveRevision(repoRoot, commitRef);
    if (!sha) {
      console.error(`fatal: not a valid object name: '${commitRef}'`);
      return 1;
    }
  } else {
    sha = getHeadCommit(repoRoot);
    if (!sha) {
      console.error('fatal: not a valid object name: HEAD');
      return 1;
    }
  }

  createTag(repoRoot, tagName, sha);
  return 0;
}

function createAnnotatedTag(repoRoot: string, tagName: string, commitRef: string | undefined, message: string): number {
  if (tagExists(repoRoot, tagName)) {
    console.error(`fatal: tag '${tagName}' already exists`);
    return 1;
  }

  let sha: string | null;
  if (commitRef) {
    sha = resolveRevision(repoRoot, commitRef);
    if (!sha) {
      console.error(`fatal: not a valid object name: '${commitRef}'`);
      return 1;
    }
  } else {
    sha = getHeadCommit(repoRoot);
    if (!sha) {
      console.error('fatal: not a valid object name: HEAD');
      return 1;
    }
  }

  const tagger = getAuthorInfo();

  const tagInfo: TagInfo = {
    object: sha,
    type: 'commit',
    tag: tagName,
    tagger: tagger.name,
    taggerEmail: tagger.email,
    taggerTimestamp: tagger.timestamp,
    taggerTz: tagger.tz,
    message,
  };

  const tagContent = createTagContent(tagInfo);
  const tagSha = writeObject(repoRoot, tagContent);

  createTag(repoRoot, tagName, tagSha);
  return 0;
}

function deleteTagCmd(repoRoot: string, tagName: string): number {
  if (!tagExists(repoRoot, tagName)) {
    console.error(`error: tag '${tagName}' not found`);
    return 1;
  }

  deleteTag(repoRoot, tagName);
  console.log(`Deleted tag '${tagName}'`);
  return 0;
}
